cmd/vault-login-oauth/vault-login-oauth.go

```diff
@@ -0,0 +1,229 @@
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/command/token"
+)
+
+type oidcLogin struct {
+	server       *http.Server
+	listener     net.Listener
+	vault        *api.Logical
+	authReqURL   string
+	redirectURL  string
+	stateNonce   string
+	authDone     context.CancelFunc
+	authErr      error
+	authResponse *api.Secret
+	role         string
+	pluginPath   string
+}
+
+func newLogin(pluginPath string, role string) *oidcLogin {
+	login := &oidcLogin{
+		server:     &http.Server{},
+		role:       role,
+		pluginPath: pluginPath,
+	}
+
+	// The callback URI must be unique for each authorization server. Since
+	// each mount of the plugin corresponds to an authorization server,
+	// incorporating the pluginPath accomplishes that.
+	//
+	// https://tools.ietf.org/html/rfc8252#section-8.10
+	callbackPath := pluginPath + "/cb"
+
+	// Create our vault client.
+	{
+		vaultClient, err := api.NewClient(nil)
+		if err != nil {
+			fatal("Couldn't create Vault client:", err)
+		}
+		login.vault = vaultClient.Logical()
+	}
+
+	// Start listening on some random localhost port, at which point we'll know
+	// what redirect_uri to give the authorization server.
+	{
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		if err != nil {
+			fatal(err)
+		}
+		login.listener = listener
+		login.redirectURL = fmt.Sprintf("http://%s%s", login.listener.Addr(), callbackPath)
+	}
+
+	// Determine the authorization request URL.
+	{
+		secret, err := login.vault.Read(pluginPath + "/auth-request")
+		if err != nil || secret == nil || secret.Data == nil {
+			fatal("Failed to get authorization request URL:", err)
+		}
+		rawURL := secret.Data["url"].(string)
+		parsedURL, err := url.Parse(rawURL)
+		if err != nil {
+			fatal("Failed parsing authorization request URL from Vault:", err)
+		}
+
+		login.stateNonce = generateNonce()
+
+		query := parsedURL.Query()
+		query.Set("redirect_uri", login.redirectURL)
+		query.Set("state", login.stateNonce)
+		parsedURL.RawQuery = query.Encode()
+		login.authReqURL = parsedURL.String()
+	}
+
+	http.HandleFunc(callbackPath, login.handleOAuthCallback)
+
+	return login
+}
+
+// HTTP handler for the redirect_uri. Once the user has gone through the
+// authentication flow at the identity provider they will be redirected here,
+// at which point we obtain the authorization code if authentication was
+// successful, or an error otherwise.
+func (login *oidcLogin) handleOAuthCallback(writer http.ResponseWriter, req *http.Request) {
+	writer.Header().Set("Content-Type", "text/plain; charset=utf-8")
+
+	query := req.URL.Query()
+	code := query.Get("code")
+	state := query.Get("state")
+
+	if state != login.stateNonce {
```

Comment:
State is recommended, but not required. Might want to accept or allow for configuration to specify nil is ok.

Reply:
I believe the authorization server MUST return it if it's specified in the authorization request, which we always do (around line 94: `query.Set("state", login.stateNonce)`) So for it to not set here would be either a buggy authorization auth server, or a CSRF attack right?

Reply:
https://tools.ietf.org/html/rfc6819#section-5.3.5
https://tools.ietf.org/html/rfc6749#section-10.12
https://tools.ietf.org/html/rfc6749#section-4.1.2

Reply:
yep, there it is 4.1.2. So many RFC's for the OAuth dance 😅